### PR TITLE
feat(router): add request latency and count metrics for router deliveries

### PR DESCRIPTION
**Pull request by koladilip**
# Description

This PR implements the `transformer_outgoing_request_latency` metric in the router to match transformer's metric structure and improve observability of HTTP requests to destination endpoints.

## Changes Made

- Added `recordTransformerOutgoingRequestLatency` method to worker struct
- Records metric only when `EndpointPath` is present in `PostParametersT`
- Includes comprehensive labels: feature, destType, endpointPath, requestMethod, module, workspaceId, destinationId, sourceId
- Added comprehensive unit tests covering various scenarios including edge cases
- Matches transformer's metric structure for consistent observability

## Implementation Details

- **Metric Name**: `transformer_outgoing_request_latency`
- **Type**: Histogram (TimerType in Go stats)
- **Location**: `router/worker.go` around line 520 for non-proxy requests
- **Condition**: Only emits when `EndpointPath` is not empty
- **Labels**: All labels match transformer's label set exactly

## Testing

- Added comprehensive unit tests with 3 test cases:
  1. Complete data with endpoint path (should emit)
  2. Empty endpoint path (should not emit)
  3. Empty job metadata array with endpoint path (should emit)
- All tests pass successfully
- No regression in existing router functionality

## Linear Ticket

https://linear.app/rudderstack/issue/INT-3991/endpoint-metrics-for-router-deliveries

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: router/worker.go [hunk 1]
```diff
@@ -1198,3 +1202,36 @@ func (w *worker) countTransformedJobStatuses(transformType string, transformedJo
...
+	}
+
+	// Get workspace and source IDs from job metadata
+	var workspaceID, sourceID string
+	if len(destinationJob.JobMetadataArray) > 0 {
+		workspaceID = destinationJob.JobMetadataArray[0].WorkspaceID
+		sourceID = destinationJob.JobMetadataArray[0].SourceID
+	}
+
+	labels := stats.Tags{
```
Comment (vinayteki95) [inline router/worker.go]: I think we should also include the other metric
`request_count`

+ did we also have statusCode label in transformer? if so, we should include that.

Comment (koladilip) [reply router/worker.go]: statusCode for Latency may not be required right? request_count one status code make sense.

**Review by vinayteki95 — APPROVED** [review]

File: router/worker.go [hunk 2]
```diff
@@ -1198,3 +1202,34 @@ func (w *worker) countTransformedJobStatuses(transformType string, transformedJo
...
+		"endpointPath":  postParams.EndpointPath,
+		"statusCode":    strconv.Itoa(resp.StatusCode),
+		"requestMethod": postParams.RequestMethod,
+		"module":        "router",
+		"workspaceId":   destinationJob.Destination.WorkspaceID,
+		"destinationId": destinationJob.Destination.ID,
+	}
+
+	// Record latency metric
+	stats.Default.NewTaggedStat("transformer_outgoing_request_latency", stats.TimerType, labels).SendTiming(duration)
```
Comment (atzoum) [inline router/worker.go]: We should cache these stats

Comment (koladilip) [reply router/worker.go]: the label set is bigger with http status codes etc so still you recommend to cache?

Comment (atzoum) [reply router/worker.go]: Yes, the timers are going to be cached in the stats library anyway already...

Comment (koladilip) [reply router/worker.go]: updated

File: router/worker.go [hunk 3]
```diff
@@ -1198,3 +1207,78 @@ func (w *worker) countTransformedJobStatuses(transformType string, transformedJo
...
+	RequestMethod string `mapstructure:"requestMethod"`
+	Module        string `mapstructure:"module"`
+	WorkspaceID   string `mapstructure:"workspaceId"`
+	DestinationID string `mapstructure:"destinationId"`
+}
+
+// ToStatsTags converts metricCacheKey to stats.Tags using mapstructure
+func (w *worker) convertDeliveryMetricLabelToStatsTags(labels deliveryMetricLabel) stats.Tags {
+	var tags map[string]string
+	err := mapstructure.Decode(labels, &tags)
```
Comment (atzoum) [inline router/worker.go]: This uses reflection, i.e. it is unnecessarily slow. We can simply do the mapping manually
```suggestion
func (dml *deliveryMetricLabel) ToStatsTags() stats.Tags {
```

File: router/worker.go [hunk 4]
```diff
@@ -1198,3 +1207,78 @@ func (w *worker) countTransformedJobStatuses(transformType string, transformedJo
...
+// getOrCreateLatencyStat returns a cached latency stat or creates a new one
+func (w *worker) getOrCreateLatencyStat(labels deliveryMetricLabel) stats.Measurement {
+	if stat, exists := w.deliveryLatencyStatsCache[labels]; exists {
+		return stat
+	}
+
+	stat := stats.Default.NewTaggedStat("transformer_outgoing_request_latency", stats.TimerType, w.convertDeliveryMetricLabelToStatsTags(labels))
+	w.deliveryLatencyStatsCache[labels] = stat
+	return stat
+}
```
Comment (atzoum) [inline router/worker.go]: Can we create a struct for this purpose in a utility package (make it goroutine-safe as well), e.g.
```go
func NewStatsCache(T key, producer func(T) stats.Measurement) StatsCache[T] {
  ...
}

type statsCache[T] struct {
  producer func(T) stats.Measurement
  cache map[T]stats.Measurement
}

func (sc *statsCache) get(T key) stats.Measurement {
  ...
}
```

Comment (koladilip) [reply router/worker.go]: done

File: router/worker.go (hunk 3 above)
Comment (koladilip) [reply router/worker.go]: Done

File: router/worker_test.go [hunk 5]
```diff
@@ -36,6 +40,21 @@ import (
...
+			destType: destType,
+		},
+		deliveryLatencyStatsCache: cache.NewStatsCache(func(labels deliveryMetricLabels) stats.Measurement {
+			return stats.Default.NewTaggedStat("transformer_outgoing_request_latency", stats.TimerType, labels.ToStatTags())
+		}),
+		deliveryCountStatsCache: cache.NewStatsCache(func(labels deliveryMetricLabels) stats.Measurement {
+			return stats.Default.NewTaggedStat("transformer_outgoing_request_count", stats.CountType, labels.ToStatTags())
+		}),
+	}
+}
```
Comment (atzoum) [inline router/worker_test.go]:
```suggestion
func createTestWorker(destType string, stat stats.Stats) *worker {
	return &worker{
		rt: &Handle{
			destType: destType,
		},
		deliveryLatencyStatsCache: cache.NewStatsCache(func(labels deliveryMetricLabels) stats.Measurement {
			return stat.NewTaggedStat("transformer_outgoing_request_latency", stats.TimerType, labels.ToStatTags())
		}),
		deliveryCountStatsCache: cache.NewStatsCache(func(labels deliveryMetricLabels) stats.Measurement {
			return stat.NewTaggedStat("transformer_outgoing_request_count", stats.CountType, labels.ToStatTags())
		}),
	}
}
```

File: router/worker_test.go [hunk 6]
```diff
@@ -629,3 +648,311 @@ func TestTransformForNonOAuthDestination(t *testing.T) {
...
+			"destinationId": "dest-456",
+		}
+
+		result := labels.ToStatTags()
+		require.Equal(t, expectedTags, result)
+	})
+
+	// Test caching behavior
+	t.Run("caching behavior", func(t *testing.T) {
+		worker := createTestWorker("TEST_DEST")
```
Comment (atzoum) [inline router/worker_test.go]:
```suggestion
		stat, err := memstats.New()
		require.NoError(t, err)
		worker := createTestWorker("TEST_DEST", stat)
```

File: router/worker_test.go [hunk 7]
```diff
@@ -629,3 +648,311 @@ func TestTransformForNonOAuthDestination(t *testing.T) {
...
+
+			// Create worker with mock router
+			worker := createTestWorker("TEST_DEST")
+
+			// Override the stats.Default to use our memstats store
+			originalStats := stats.Default
+			stats.Default = statsStore
+			defer func() {
+				stats.Default = originalStats
+			}()
```
Comment (atzoum) [inline router/worker_test.go]:
```suggestion
			stat, err := memstats.New()
			require.NoError(t, err)

			// Create worker with mock router
			worker := createTestWorker("TEST_DEST", stat)
```

File: router/worker_test.go [hunk 8]
```diff
@@ -629,3 +648,311 @@ func TestTransformForNonOAuthDestination(t *testing.T) {
...
+
+			// Call the method under test
+			worker.recordTransformerOutgoingRequestMetrics(tc.postParams, tc.destinationJob, tc.resp, tc.duration)
+
+			// Get all recorded metrics
+			allMetrics := statsStore.GetAll()
```
Comment (atzoum) [inline router/worker_test.go]:
```suggestion
			allMetrics := stat.GetAll()
```

File: router/worker_test.go (hunk 5 above)
Comment (koladilip) [reply router/worker_test.go]: addressed

File: router/worker_test.go (hunk 7 above)
Comment (koladilip) [reply router/worker_test.go]: addressed

File: router/worker_test.go (hunk 8 above)
Comment (koladilip) [reply router/worker_test.go]: addressed

File: router/worker_test.go (hunk 6 above)
Comment (koladilip) [reply router/worker_test.go]: addressed
